activerecord/lib/active_record/association_relation.rb

```diff
@@ -37,6 +37,18 @@ def create!(attributes = nil, &block)
       end
     end
 
+    %w(insert insert_all insert! insert_all! upsert upsert_all).each do |method|
+      class_eval <<~RUBY
+        def #{method}(attributes, **kwargs)
+          if @association.reflection.through_reflection?
+            raise ArgumentError, "Bulk insert is currently not supported for through relations"
```

Comment:
@kamipo Can we change the error message to say `Bulk insert or update` since `upsert` perform updates? Also, `through relations` -> `has_many :through association` if that looks proper to you.

```suggestion
            raise ArgumentError, "Bulk insert or update is currently not supported for has_many :through association"
```

Reply:
Also, is `ArgumentError` correct here? We haven't implemented bulk insert for has_many through association. Shouldn't it be `NotImplementedError`?

Reply:
AFAIK `NotImplementedError` is not meant to be raised for missing features, but for missing platform features. https://ruby-doc.org/core-2.5.0/NotImplementedError.html